### ISSUE TITLE
Restart a partial download

### ISSUE THREAD

**Issue by LattaruoloAndrea** [SDFIdk/CDSETool#69]
## Summary
Hello, since the data that we download it can be demanding it would be a great idea to allow restarting partial dowloads. 

### Basic example 
Something like the code below could work 
For example a file it's been stopped with M bytes downloaded, we could do a request on dowload starting from M+1 bytes
```Python
        already_downloaded_bytes = 0
        headers = {"Authorization": f"Bearer {access_token}"}
        if os.exists(relative_path):
            already_downloaded_bytes =relative_path.stat().st_size # Get the M bytes already donwloaded
            headers = {"Authorization": f"Bearer {access_token}","Range": "bytes={}-".format(already_downloaded_bytes)}
``` 
The only issue that I have found is that every time I tried to use the header with bytes that was different from 0 the api gave me and error. I don't know if it's possible to achive partial downloads but it would be useful. 

###Where to save the partial files
I have seen that the partial files during the dowload are created with 
```Python
fd, tmp = tempfile.mkstemp()
```
I guess it would be better to create a partial file in the path that we insert adding a '.incomplete' to the name of the file 
```Python
def download_feature(feature, path, options=None):
    """
    Download a single feature

    Returns the feature ID
    """
    options = options or {}
    url = _get_feature_url(feature)
    filename = feature.get("properties").get("title")

    if not url or not filename:
        return feature.get("id")

    overwrite = options.get("overwrite_existing", False)
    continue_ = options.get("continue_existing", True)

    if not os.path.exists(Path(path)):
        os.makedirs(path)
    result_path = os.path.join(path, filename.replace(".SAFE", ".zip"))
    temp_path =  Path(path) / filename.replace(".SAFE", ".zip.incomplete") #os.path.join(path, filename.replace(".SAFE", ".zip.incomplete"))
    if not overwrite and os.path.exists(result_path):
        return feature.get("id")
    
    if not continue_ and os.path.exists(temp_path):
        #TODO decide what to do with partlial files, 
        """
        it should be appropriate to restart the download from the
        from where it let off istead of the beginning.
        The current version remove the partial file and restart the download
        from the beginning.
        """
        return feature.get("id")
    if os.path.exists(temp_path):
        os.remove(temp_path)
```
like this is the implementation that I made of where to dave the partial downloads on the download_feature() function

**Comment by pedep** [comment]
Hi @LattaruoloAndrea,

Thank you for your feature request :1st_place_medal:

> The only issue that I have found is that every time I tried to use the header with bytes that was different from 0 the api gave me and error. I don't know if it's possible to achive partial downloads but it would be useful.

I have had a play around with this, and it seems to be a bit buggy.
I got range start working, but it seems to ignore the ranges end argument

Heres a bit of junk code that got me some results
```python
headers = {"Authorization": f"Bearer {access_token}"}

session = requests.Session()
session.headers.update(headers)

# this is required to get the actual download url. copernicus has some weird redirect jank
from cdsetool.download import  _follow_redirect
url = _follow_redirect("https://catalogue.dataspace.copernicus.eu/download/c2f31545-e386-57dc-9762-602c275c724b", session)

headers = {"Authorization": f"Bearer {access_token}","Range": "bytes=827036973-827037973"}
session = requests.Session()
session.headers.update(headers)

r = session.get(url, stream=True)

for k,v in r.headers.items():
    print(f"{k}: {v}")
print(r.status_code)

content_length = 0
for chunk in r.iter_content(chunk_size=1024):
    content_length += len(chunk)
    print(".", end="", flush=True)

print("")
print(content_length)
```
The `content_length` printed at the end does not match the expected `|827036973-827037973| = 1000`, but rather the 2100 bytes remaining

The `content-range` header also seems to contain invalid data, as the requested range is displayed as `content-range: bytes 2100-827039073/827039073` rather than the expected `content-range: bytes 827036973-827037973/827039073`

While its a bit problematic, i think this might be usable. 
I have not yet tested if the result returned by the range request matches the same offset in the full file.
Even with the bugs this should be usable, assuming no corruption in the returned data.

> I guess it would be better to create a partial file in the path that we insert adding a '.incomplete' to the name of the file

I like this approach, as i have suggested myself in https://github.com/SDFIdk/CDSETool/issues/51, but as suggested in
https://github.com/SDFIdk/CDSETool/issues/51#issuecomment-1845107761 there might be some security concerns with this approach.

Having thought about it for a while, i think using the filename based tempfiles should be an acceptable risk. We have a reasonable basis for treating CDSE as a secure source of files, as long as we do not disable certificate validation in `requests`.

Other software uses a similar approach:
Chrome https://github.com/chromium/chromium/blob/main/chrome/browser/download/download_target_determiner.cc#L1074
https://github.com/chromium/chromium/blob/2e0cc3ea8bdf51861de5f0fbd2134f49eb18a46a/chrome/browser/download/download_target_determiner.h#L115
Rsync
https://github.com/WayneD/rsync/blob/2f9b963abaa52e44891180fe6c0d1c2219f6686d/receiver.c#L90

A different approach to tracking incomplete files might be achieved by placing a small database in the download directory. Something simple like JSON or pickle, just to keep track of in-progress files.

The solution mentioned here might be the easiest, as it allows users to customize the tmpdir while allowing us to implement an opinionated default.
https://github.com/SDFIdk/CDSETool/issues/51#issuecomment-1845366976

**Comment by LattaruoloAndrea** [comment]
Hello @pedep, thanks for the response. The tempfiles directory should work, do we still wanna implemnet the ".inclomplete" format since a user could use the same forlder for the download direcotry and the tmpdir.
```Python
list(
    download_features(
        features,
        "use_same_path_to_dir",
        {
            "tmpdir": "use_same_path_to_dir",
            "concurrency": 4,
            "monitor": StatusMonitor(),
            "credentials": Credentials("username", "password"),
        },
    )
)
```
And if tmpdir is not specified we could just use the download_dir or will we use the. (which personally I don't like)
```Python
fd, tmp = tempfile.mkstemp()
```
Regarding the partial downloads I was asking because I did have problems as well, from the example that you showed I tried the same but not with the range of bytes but just from the beginning of the bytes
```
from cdsetool.download import  _follow_redirect
url = _follow_redirect("https://catalogue.dataspace.copernicus.eu/download/c2f31545-e386-57dc-9762-602c275c724b", session)
already_downloaded_bytes = 827036973
headers = {"Authorization": f"Bearer {access_token}","Range": "bytes={}-".format(already_downloaded_bytes)}
session = requests.Session()
session.headers.update(headers)
```
Which would be our case, since we can check the already_dowloaded_bytes from a partial dowloads and then start from that point. Regarding finding the incomplete files I did already made the procedures that find those files something like 
```Pyhton 
def download_feature(feature, path, options=None):
   ...
   temp_path =  Path(path) / filename.replace(".SAFE", ".zip.incomplete")
   if os.path.exists(result_path):
       # logic that we want to implement -> maybe continue a partial file
    else:
       # full file to download
```

**Comment by pedep** [comment]
> do we still wanna implemnet the ".inclomplete" format since a user could use the same forlder for the download direcotry and the tmpdir.

Yes, i think this would be a good idea. I think prepending a period like rsync does, and appending `.incomplete` would be my preferred solution.

> And if tmpdir is not specified we could just use the download_dir or will we use the. (which personally I don't like)

The download dir, or possibly a directory inside (eg. `/path/to/download_dir/.tmp/`) would be a fine default, while allowing users to specify a custom dir. Rsync defaults to using the download dir too.
This solution would close this issue too https://github.com/SDFIdk/CDSETool/issues/51
